### PR TITLE
Cleanup Api docs

### PR DESCRIPTION
**Pull request by SpacemanPaul**
### Reason for this pull request

Cleanup and expansion of the curated API docs.

### Proposed changes

- Improve index API documentation,
- Fix various autodoc and autosummary issues.
- Add `LineageTree` to model documentation
- Replace lingering reference to `DatasetType` with `Product`
- Fix broken directives in `masking`

 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

----
📚 Documentation preview 📚: https://datacube-core--1688.org.readthedocs.build/en/1688/

### CODE REVIEW

File: docs/api/core-classes/lineage-tree.rst [hunk 1]
```diff
@@ -0,0 +1,17 @@
+==========
+LinageTree
```
Comment (Kirill888) [inline docs/api/core-classes/lineage-tree.rst]: spelling

File: docs/api/index.rst [hunk 2]
```diff
@@ -15,6 +15,7 @@ The following pages provide a full API reference for the ``datacube`` python lib
     core-classes/geometry
     core-classes/measurement
     core-classes/metadataType
+    core-classes/lineage-tree
```
Comment (Ariana-B) [inline docs/api/index.rst:18]: Minor, but would be nice for this to use camel case to be consistent with `metadataType` (or the other way around)

File: docs/installation/extending-odc.rst [hunk 3]
```diff
@@ -215,24 +171,34 @@ to connect to an ``Index``. This PR extends this with an
...
 is expected to supply implementations of all these abstract base classes. If any abstract
 methods is not relevant to or implementable by a particular Index Driver, that method should
 defined to raise a ``NotImplementedError``.
 
-Default Implementation
+Legacy Implementation
+---------------------
+
+The legacy ``postgres`` index driver uses a PostgreSQL database for all storage and
+retrieval, with search support by json indexes.
```
Comment (Ariana-B) [inline docs/installation/extending-odc.rst]:
```suggestion
retrieval, with search supported by json indexes.
```

File: docs/installation/extending-odc.rst [hunk 4]
```diff
@@ -215,24 +171,34 @@ to connect to an ``Index``. This PR extends this with an
...
-retrieval.
+The new ``postgis`` index driver uses a PostgreSQL database for all storage and
+retrieval, with spatial search using postgis spatial indexes.
 
-If explicitly declaring the default index driver, you can use either ``default`` or ``postgres``
-as the ``index_driver`` in the configuration file.
+Default Implementation
+----------------------
+
+Not specifying and index driver (or specifying ``default`` as the index driver) results in
```
Comment (Ariana-B) [inline docs/installation/extending-odc.rst]:
```suggestion
Not specifying an index driver (or specifying ``default`` as the index driver) results in
```

**Review by Ariana-B — APPROVED** [review]

**Pull request merged**